### PR TITLE
Custom expiration time

### PR DESCRIPTION
**Pull request by achiyahb**
1. Save authenticated user's links
2. Enable custom expression time
3. Docs
4. Units tests
experation time feature diagram

### CODE REVIEW

File: backend/prisma/schema.prisma [hunk 1]
```diff
@@ -20,6 +20,18 @@ model User {
   verified          Boolean @default(false) // Is the mail verification
   verificationToken String? // Verification token of the mail
   refreshToken      String?
+  shortsUrls        ShortUrl[]
+}
+
+model ShortUrl {
```
Comment (origranot) [inline backend/prisma/schema.prisma:26]: Change ShortUrl to Url

File: backend/prisma/schema.prisma [hunk 2]
```diff
@@ -20,6 +20,18 @@ model User {
...
+  id                String  @id @default(uuid())
+  shortUrl          String  @unique
```
Comment (origranot) [inline backend/prisma/schema.prisma:28]: Change to shortenedUrl

File: backend/src/shortener/shortener.controller.ts [hunk 3]
```diff
@@ -12,34 +24,25 @@ export class ShortenerController {
   @Get(':shortUrl')
   async findOne(@Param('shortUrl') shortUrl: string) {
     const originalUrl = await this.shortenerService.getOriginalUrl(shortUrl);
-    if (!originalUrl) {
+    if (originalUrl) {
+      return originalUrl;
+    }
+    const premiumUrl = await this.shortenerService.getPremiumUrl(shortUrl);
+    if (!premiumUrl) {
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.ts:31]: ```getPremiumUrl``` should be inside the logic of ```getOriginalUrl```

File: backend/src/shortener/shortener.controller.spec.ts [hunk 4]
```diff
@@ -27,64 +40,82 @@ describe('ShortenerController', () => {
...
       expect(short).toStrictEqual({ newUrl: 'best' });
     });
 
     it('should throw an error of invalid url', () => {
-      const body: ShortenerDTO = { originalUrl: 'invalid-url' };
+      const body: ShortenerDto = { originalUrl: 'invalid-url' };
+      const req = { user: { id: 'user_id' } } as any as Request;
       expect(async () => {
-        await shortenerController.shortener(body);
+        await shortenerController.shortener(body, req);
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:58]: What will happen if you don't send request object? I think we should test this case also

File: backend/src/shortener/shortener.controller.spec.ts [hunk 5]
```diff
@@ -27,64 +40,82 @@ describe('ShortenerController', () => {
...
+      const body: ShortenerDto = { originalUrl: 'https://github.com/origranot/reduced.to' };
+      const req = { user: { id: 'user_id' } } as any as Request;
+
+      try {
+        await shortenerController.shortener(body, req);
+        throw new Error('Expected an error to be thrown!');
+      } catch (err) {
+        expect(err.message).toBe('The URL is already shortened...');
+      }
+    });
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:73]: I guess I did it and you just added the req object. this is a weird test tbh.
Just to keep in my mind to change this.

File: backend/src/shortener/shortener.controller.spec.ts [hunk 6]
```diff
@@ -27,64 +40,82 @@ describe('ShortenerController', () => {
...
+      jest
+        .spyOn(shortenerService, 'addUrl')
+        .mockRejectedValue(new Error('Error adding URL to the database'));
+      const body: ShortenerDto = { originalUrl: 'https://github.com/origranot/reduced.to' };
+      const req = { user: { id: 'user_id' } } as any as Request;
+      expect(async () => {
+        await shortenerController.shortener(body, req);
+      }).rejects.toThrow();
+    });
+  });
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:87]: Again, test it without request object.

File: backend/src/shortener/shortener.service.spec.ts [hunk 7]
```diff
@@ -93,24 +95,93 @@ describe('ShortenerService', () => {
...
       const url = 'https://github.com/origranot/reduced.to';
 
-      const result = service.isUrlAlreadyShortend(url);
+      const result = service.isUrlAlreadyShortened(url);
       expect(result).toBe(false);
     });
   });
+
+  describe('getUrlFromDb', () => {
+    it('should return the premium url', async () => {
```
Comment (origranot) [inline backend/src/shortener/shortener.service.spec.ts:120]: Remove "premium" word. in all the files

File: backend/src/shortener/shortener.service.spec.ts [hunk 8]
```diff
@@ -93,24 +95,93 @@ describe('ShortenerService', () => {
...
+      prisma.url.findFirst = jest.fn().mockReturnValueOnce({
+        originalUrl: 'original_url',
+      });
+      const result = await service.getUrlFromDb('good_url');
+      expect(result).toBe('original_url');
+    });
+
+    it('should return null if the premium url is not found', async () => {
+      prisma.url.findMany = jest.fn().mockReturnValueOnce(undefined);
```
Comment (origranot) [inline backend/src/shortener/shortener.service.spec.ts:129]: Why do you use findMany? we should always use findOne..

File: backend/src/shortener/shortener.service.spec.ts [hunk 9]
```diff
@@ -93,24 +95,93 @@ describe('ShortenerService', () => {
...
+      jest.spyOn(service, 'isShortenedUrlAvailable').mockResolvedValue(true);
+      jest.spyOn(service, 'addUrl').mockResolvedValue(undefined);
+      jest.spyOn(service, 'isUrlAlreadyShortened').mockReturnValue(false);
+
+      const body: ShortenerDto = { originalUrl: 'https://github.com/origranot/reduced.to' };
+      const short = await service.createShortUrl(body);
+      expect(short).toStrictEqual({ newUrl: 'best' });
+    });
+
+    it('should throw an error of invalid url', () => {
```
Comment (origranot) [inline backend/src/shortener/shortener.service.spec.ts:147]: Wrap this with describe of createShortUrl

File: backend/src/shortener/shortener.service.ts [hunk 10]
```diff
@@ -33,30 +50,98 @@ export class ShortenerService {
...
   /**
    * Add the short url to the server routes.
    * @param {String} originalUrl The original url.
-   * @param {String} shortUrl The shorten url.
+   * @param {String} shortenedUrl The shorten url.
+   */
+  addUrl = async (originalUrl: string, shortenedUrl: string) => {
+    const isShortenedUrlAvailable = await this.isShortenedUrlAvailable(shortenedUrl);
+    if (!isShortenedUrlAvailable) {
+      throw new Error('ShortenedURL already taken');
```
Comment (origranot) [inline backend/src/shortener/shortener.service.ts:75]: Change the error message to be easier to read - 
Shortened Url is already taken

File: backend/src/shortener/shortener.service.ts [hunk 11]
```diff
@@ -33,30 +50,98 @@ export class ShortenerService {
...
+  /**
+   * Create a short URL based on the provided data.
+   * @param {ShortenerDto} body The data for creating a short URL.
+   * @returns {Promise<{ newUrl: string }>} Returns an object containing the newly created short URL.
    */
-  addUrl = async (originalUrl: string, shortUrl: string) => {
-    const isShortUrlAvailable = await this.isShortUrlAvailable(shortUrl);
-    if (!isShortUrlAvailable) {
-      throw new Error('Short URL already taken');
+  createShortUrl = async (
```
Comment (origranot) [inline backend/src/shortener/shortener.service.ts:85]: Should we change all of "shortUrl" to "shortenedUrl" ?

File: backend/src/shortener/shortener.service.ts [hunk 12]
```diff
@@ -33,30 +50,98 @@ export class ShortenerService {
...
+   * @returns {Promise<string|null>} - The original URL if the original URL is found and valid, otherwise null.
+   */
+  async getUrlFromDb(shortenedUrl: string) {
+    const url = await this.prisma.url.findFirst({
+      where: {
+        shortenedUrl,
+      },
+    });
+    if (url?.expirationTime && url.expirationTime < new Date()) {
+      return null;
```
Comment (origranot) [inline backend/src/shortener/shortener.service.ts:143]: Maybe we should also remove it from the database?

**Review by origranot — CHANGES_REQUESTED** [review]
Great job, hopefully we can push this asap 💯

File: backend/src/cache/cache.service.ts [hunk 13]
```diff
@@ -1,12 +1,37 @@
...
+    }
+    return (expirationTimeAsNumber - now) / 1000;
+  }
+
+  getSmallerTtl(customTtl: number) {
+    const defaultTtl = this.appConfigService.getConfig().cache.ttl;
+    if (defaultTtl) {
+      return Math.min(customTtl, defaultTtl);
+    } else {
+      return customTtl;
```
Comment (origranot) [inline backend/src/cache/cache.service.ts:34]: defaultTtl can't be undefined.
let's do something like:
```suggestion
return Math.min(ttl, this.appConfigService.getConfig().cache.ttl);

File: backend/src/config/config.factory.ts [hunk 14]
```diff
@@ -75,12 +78,17 @@ export interface NovuConfig {
...
+  ttl: number;
+}
+
 export interface Configuration {
   app: AppConfig;
   logger: LoggerConfig;
   front: FrontConfig;
   rateLimit: RateLimitConfig;
   redis: RedisConfig;
+  cache: CacheConfig;
```
Comment (origranot) [inline backend/src/config/config.factory.ts:91]: Not needed

File: backend/src/shortener/shortener.service.ts (hunk 10 above)
Comment (achiyahb) [reply backend/src/shortener/shortener.service.ts:76]: done

File: backend/src/shortener/shortener.service.ts (hunk 11 above)
Comment (achiyahb) [reply backend/src/shortener/shortener.service.ts:93]: done

File: backend/src/cache/cache.service.ts (hunk 13 above)
Comment (achiyahb) [reply backend/src/cache/cache.service.ts]: done

File: backend/src/shortener/shortener.service.spec.ts (hunk 8 above)
Comment (achiyahb) [reply backend/src/shortener/shortener.service.spec.ts]: done

File: backend/src/shortener/shortener.service.spec.ts (hunk 9 above)
Comment (achiyahb) [reply backend/src/shortener/shortener.service.spec.ts:151]: done

File: backend/src/config/config.factory.ts [hunk 15]
```diff
@@ -34,6 +34,9 @@ export const configFactory: ConfigFactory<{ config: Configuration }> = () => {
       novu: {
         apiKey: process.env.NOVU_API_KEY,
       },
+      cache: {
+        ttl: +process.env.CACHE_DEFAULT_TTL || 60 * 60 * 24 * 7, // Seconds
+      },
```
[the remark this answers is not in the document]
Comment (achiyahb) [reply backend/src/config/config.factory.ts]: done

File: backend/src/config/config.factory.ts (hunk 14 above)
Comment (achiyahb) [reply backend/src/config/config.factory.ts]: done

File: backend/src/shortener/shortener.controller.spec.ts (hunk 4 above)
Comment (achiyahb) [reply backend/src/shortener/shortener.controller.spec.ts]: done

File: backend/src/auth/utils/ttl.util.spec.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+  });
+
+  describe('convertExpirationTimeToTtl', () => {
+    it('should return null if expirationTime has falsy value', () => {
+      const expirationTime = null;
+      const ttl = ttlUtil.convertExpirationTimeToTtl(expirationTime);
+      expect(ttl).toBeNull();
+    });
+
+    it('should return null if expiration time is smaller than now', () => {
```
Comment (origranot) [inline backend/src/auth/utils/ttl.util.spec.ts:31]: earlier

File: backend/src/auth/utils/ttl.util.spec.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (origranot) [inline backend/src/auth/utils/ttl.util.spec.ts:1]: Move it to cache/utils

File: backend/src/auth/utils/ttl.util.spec.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    it('should return null if expiration time is smaller than now', () => {
+      const expirationTime = '2021-01-01 00:00:00';
+      const ttl = ttlUtil.convertExpirationTimeToTtl(expirationTime);
+      expect(ttl).toBeNull();
+    });
+
+    it('should return ttl if expiration time is bigger than now', () => {
+      const date = new Date().setTime(new Date().getTime() + 1000 * 60 * 60 * 24 * 390);
+      const year = new Date(date).getFullYear();
+      const expirationTime = `${year}-01-01 00:00:00`;
```
Comment (origranot) [inline backend/src/auth/utils/ttl.util.spec.ts:40]: Just do newDate() * 100 or something

File: backend/src/auth/utils/ttl.util.ts [hunk 19]
```diff
@@ -0,0 +1,23 @@
+import { Injectable } from '@nestjs/common';
+import { AppConfigService } from '../../config/config.service';
+
+@Injectable()
+export class TtlUtil {
+  constructor(private readonly appConfigService: AppConfigService) {}
+
+  convertExpirationTimeToTtl(expirationTime: string): number {
```
Comment (origranot) [inline backend/src/auth/utils/ttl.util.ts:8]: I think this method should get Date object.

File: backend/src/auth/utils/ttl.util.ts [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    const now = new Date().getTime();
+    if (expirationTimeAsNumber < now) {
+      return null;
+    }
+    return (expirationTimeAsNumber - now) / 1000;
+  }
+
+  getSmallerTtl(ttl: number) {
+    return Math.min(ttl, this.appConfigService.getConfig().redis.ttl);
+  }
```
Comment (origranot) [inline backend/src/auth/utils/ttl.util.ts:22]: I think we should consider remove this method and use it in the service itself (shortener service).
We can create a util class with exported functions. (remove Injectable from this class).

File: backend/src/cache/cache.module.ts [hunk 21]
```diff
@@ -25,13 +26,13 @@ import { AppCacheService } from './cache.service';
...
+          ttl: config.getConfig().redis.ttl,
         };
       },
       inject: [AppConfigService],
     }),
   ],
-  providers: [AppCacheService],
-  exports: [AppCacheService],
+  providers: [AppCacheService, TtlUtil],
+  exports: [AppCacheService, TtlUtil],
```
Comment (origranot) [inline backend/src/cache/cache.module.ts:36]: Remove TTlUtil

File: backend/src/shortener/shortener.controller.spec.ts [hunk 22]
```diff
@@ -1,10 +1,10 @@
-import { AppConfigModule } from './../config/config.module';
-import { AppCacheModule } from './../cache/cache.module';
+import { AppConfigModule } from '../config/config.module';
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:1]: Use supertest package to test with real request.

see users.controller.spec

File: backend/src/shortener/shortener.controller.spec.ts [hunk 23]
```diff
@@ -26,24 +35,48 @@ describe('ShortenerController', () => {
...
+      jest.spyOn(shortenerService, 'createUsersShortUrl').mockResolvedValue({ newUrl: 'url.com' });
+      jest.spyOn(shortenerService, 'createShortenedUrl').mockResolvedValue(null);
+
+      const body: ShortenerDto = { originalUrl: 'https://github.com/origranot/reduced.to' };
+      const req = { user: { id: 'has_id' } } as any as Request;
+      const short = await shortenerController.shortener(body, req);
+      expect(short).toStrictEqual({ newUrl: 'url.com' });
     });
 
+    it('guest user should return null when createShortenedUrl return null', async () => {
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:58]: start all the tests with "should"

File: backend/src/shortener/shortener.controller.spec.ts [hunk 24]
```diff
@@ -26,24 +35,48 @@ describe('ShortenerController', () => {
...
 
   describe('shortener', () => {
-    it('should return a shortern url', async () => {
-      jest.spyOn(shortenerService, 'generateShortUrl').mockReturnValue('best');
-      jest.spyOn(shortenerService, 'isShortUrlAvailable').mockResolvedValue(true);
-      jest.spyOn(shortenerService, 'addUrl').mockResolvedValue(undefined);
-      jest.spyOn(shortenerService, 'isUrlAlreadyShortend').mockReturnValue(false);
+    it('authenticated user should return null when createUsersShortUrl return null', async () => {
+      jest.spyOn(shortenerService, 'createUsersShortUrl').mockReturnValue(null);
+      jest.spyOn(shortenerService, 'createShortenedUrl').mockResolvedValue({ newUrl: 'url' });
```
Comment (origranot) [inline backend/src/shortener/shortener.controller.spec.ts:40]: Can we put this 2 shared lines in beforeEach function?

File: backend/src/shortener/shortener.service.spec.ts [hunk 25]
```diff
@@ -1,25 +1,49 @@
...
+          useFactory: () => ({
+            url: {
+              create: jest.fn().mockResolvedValue({
+                id: 'cdd49d7b-3ad6-41e6-8010-47887b26cb88',
+                shortenedUrl: 'sqpjf',
+                originalUrl: 'https://support.co-ing.co/notification/dispatchers',
+                userId: '26419f47-97bd-4f28-ba2d-a33c224fa4af',
+                description: null,
+                expirationTime: null,
+                createdAt: '2023-05-28T15:16:06.837Z' as any,
```
Comment (origranot) [inline backend/src/shortener/shortener.service.spec.ts:34]: Write it as a const

File: backend/src/shortener/shortener.service.spec.ts [hunk 26]
```diff
@@ -93,24 +108,176 @@ describe('ShortenerService', () => {
...
+        originalUrl: 'original_url',
+        expirationTime: new Date(Date.now() - 1000 * 60),
+      });
+      const result = await service.getUrlFromDb('expired_url');
+      expect(result).toBeNull();
+    });
+
+    it('should return url if expiration time bigger than now', async () => {
+      prisma.url.findFirst = jest.fn().mockReturnValueOnce({
+        originalUrl: 'original_url',
```
Comment (origranot) [inline backend/src/shortener/shortener.service.spec.ts:158]: use consts

**Pull request merged**